lib/logstash/pluginmanager/update.rb

```diff
@@ -1,51 +1,80 @@
 require 'clamp'
 require 'logstash/namespace'
 require 'logstash/pluginmanager/util'
+require 'logstash/pluginmanager/base'
 require 'jar-dependencies'
 require 'jar_install_post_install_hook'
 require 'file-dependencies/gem'
-
 require "logstash/gemfile"
 require "logstash/bundler"
 
-class LogStash::PluginManager::Update < Clamp::Command
+class LogStash::PluginManager::Update < LogStash::PluginManager::Base
   parameter "[PLUGIN] ...", "Plugin name(s) to upgrade to latest version"
 
   def execute
-    gemfile = LogStash::Gemfile.new(File.new(LogStash::Environment::GEMFILE_PATH, "r+")).load
-    # keep a copy of the gemset to revert on error
-    original_gemset = gemfile.gemset.copy
-
-    previous_gem_specs_map = find_latest_gem_specs
+    local_gems = gemfile.locally_installed_gems
 
-    # create list of plugins to update
-    plugins = unless plugin_list.empty?
-      not_installed = plugin_list.select{|plugin| !previous_gem_specs_map.has_key?(plugin.downcase)}
-      raise(LogStash::PluginManager::Error, "Plugin #{not_installed.join(', ')} is not installed so it cannot be updated, aborting") unless not_installed.empty?
-      plugin_list
+    if update_all? || !local_gems.empty?
+      error_plugin_that_use_path!(local_gems)
     else
-      previous_gem_specs_map.values.map{|spec| spec.name}
+      plugins_with_path = plugin_list & local_gems
+      error_plugin_that_use_path!(plugins_with_path) if plugins_with_path.size > 0 
     end
 
+    update_gems!
+  end
+
+  private
+  def error_plugin_that_use_path!(plugins)
+    signal_error("You have installed plugins from a .gem or you have manually defined a plugin in the Gemfile, we cannot update all or update this specific plugin, problematic plugins: #{plugins.collect(&:name).join(",")}")
```

Comment:
Maybe something shorter like "Update is not supported for manually defined plugins or local .gem plugin installations: <list_of_plugins>"?